crates/sui-types/src/event.rs

```diff
@@ -1,22 +1,78 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use move_core_types::language_storage::StructTag;
+use move_core_types::language_storage::{ModuleId, StructTag};
+use name_variant::NamedVariant;
 use serde::{Deserialize, Serialize};
 use serde_with::{serde_as, Bytes};
 
-/// User-defined event emitted by executing Move code.
-/// Executing a transaction produces an ordered log of these
+use crate::{
+    base_types::{ObjectID, SequenceNumber, SuiAddress, TransactionDigest},
+    committee::EpochId,
+    messages_checkpoint::CheckpointSequenceNumber,
+};
+
+/// A universal Sui event type encapsulating different types of events
+#[derive(Debug, Clone, PartialEq)]
+pub struct EventEnvelope {
+    /// UTC timestamp in milliseconds since epoch (1/1/1970)
+    timestamp: u64,
+    /// Transaction digest of associated transaction, if any
+    tx_digest: Option<TransactionDigest>,
+    /// Specific event type
+    event: SuiEvent,
+}
+
+impl EventEnvelope {
+    pub fn new(timestamp: u64, tx_digest: Option<TransactionDigest>, event: SuiEvent) -> Self {
+        Self {
+            timestamp,
+            tx_digest,
+            event,
+        }
+    }
+
+    pub fn event_type(&self) -> &'static str {
+        self.event.variant_name()
+    }
+}
+
+/// Specific type of event
 #[serde_as]
-#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
-pub struct Event {
-    pub type_: StructTag,
-    #[serde_as(as = "Bytes")]
-    pub contents: Vec<u8>,
+#[derive(Eq, Debug, Clone, PartialEq, NamedVariant, Deserialize, Serialize, Hash)]
+pub enum SuiEvent {
+    /// Move-specific event
+    MoveEvent {
+        type_: StructTag,
+        #[serde_as(as = "Bytes")]
+        contents: Vec<u8>,
+    },
+    /// Module published
+    Publish { package_id: ObjectID },
+    /// Transfer coin
+    TransferCoin {
```

Comment:
I'm thinking something like this
```
pub enum DerivedEvent {
    ObjectChangelog {...},
    AddressChangelog {...},
    Publish {...}
}

pub enum SystemEvent {
    NewEpoch {...},
    NewCheckpoint {...},
}


pub enum SuiEvent {
    MoveEvent {...},
    SystemEvent(SystemEvent),
    DerivedEvent(DerivedEvent),
}
```
this makes `SuiEvent` more concise

Reply:
Hmmm, this might be fine but I'm wondering if we really need to distinguish DerivedEvents from SystemEvents.

Reply:
My understanding is `DerivedEvents` is derived from txns. `Publish` a module is through a txn, there's no Move Event emitted so we derive it from a txn. SystemEvent like `HardFork` is not done through a txn, as a comparison

Reply:
Ah... maybe this is just terminology or a way of thinking about it.  I think of a publish event not as "derived" but as coming from the Sui validator.  It's just not user generated.  I don't look at a Publish event any differently than a system event like new checkpoint, it's just something else.